File: HoTT-MdlViewer-SWT/src/main/java/de/treichels/hott/mdlviewer/swt/Launcher.java
package de.treichels.hott.mdlviewer.swt;

import de.treichels.hott.messages.Messages;
import de.treichels.hott.util.Util;
import org.jetbrains.annotations.NonNls;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.MalformedURLException;
import java.net.URISyntaxException;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.jar.Attributes;
import java.util.jar.JarFile;
import java.util.jar.Manifest;

public class Launcher {
    @NonNls
    private static final String LOG_DIR = "log.dir";
    @NonNls
    public static final String MDL_DIR = "mdl.dir";
    @NonNls
    private static final String PROGRAM_DIR = "program.dir";
    @NonNls
    public static final String PROGRAM_VERSION = "program.version";

    /**
     * Initialize SWT. Make sure that the correct SWT library is in the classpath.
     */
    private static void initSwt() throws NoSuchMethodException, SecurityException, IllegalAccessException, IllegalArgumentException, InvocationTargetException,
            MalformedURLException, ClassNotFoundException {
        try {
            // check if swt is already in the classpath
            Class.forName("org.eclipse.swt.widgets.Dialog");
        } catch (final ClassNotFoundException e) {
            // nope - we need to add it

            // determine correct swt jar
            @NonNls String osName = System.getProperty("os.name").toLowerCase();
            if (osName.startsWith("linux"))
                osName = "linux";
            else if (osName.startsWith("windows"))
                osName = "windows";
            else if (osName.startsWith("mac")) osName = "mac";

            final String osArch = System.getProperty("os.arch").toLowerCase();
            final String swtJarName = String.format("swt-%s-%s.jar", osName, osArch);
            final File swtJar = new File(new File(System.getProperty(PROGRAM_DIR), "swt"), swtJarName);
            if (!swtJar.exists() || !swtJar.isFile()) throw new ClassNotFoundException(Messages.getString("Launcher.SWTNotFound", swtJar.getAbsolutePath()));

            // add swt to classpath
            // Note: this is a hack. Not all platforms may use URLClassLoader as
            // standard. Make protected addURL Method available via reflection
            // and
            // invoke it
            final URLClassLoader classLoader = (URLClassLoader) Thread.currentThread().getContextClassLoader();
            final Method method = URLClassLoader.class.getDeclaredMethod("addURL", URL.class);
            method.setAccessible(true);
            method.invoke(classLoader, swtJar.toURI().toURL());
        }
    }

    /**
     * Initialize system properties.
     */
    public static void initSystemProperties() throws URISyntaxException, IOException {
        // get the location of this class
        final URL url = Launcher.class.getProtectionDomain().getCodeSource().getLocation();
        File source = new File(url.toURI());

        if (source.getName().endsWith(".jar") || source.getName().endsWith(".exe")) {
            // read program version from manifest
            JarFile jarFile = null;
            try {
                jarFile = new JarFile(source);
                final Manifest manifest = jarFile.getManifest();
                @NonNls final Attributes attributes = manifest.getMainAttributes();
                final String version = Messages.getString("Launcher.Version", attributes.getValue("Implementation-Version"),
                        attributes.getValue("Implementation-Build"));
                System.setProperty(Launcher.PROGRAM_VERSION, version);
            } finally {
                if (jarFile != null) jarFile.close();
            }
        } else {
            if (!System.getProperties().containsKey(PROGRAM_VERSION)) System.setProperty(Launcher.PROGRAM_VERSION, Messages.getString("Launcher.Unknown"));

            // application was packaged as individual class files, find the
            // classes
            // directory
            while (!source.getName().equals("classes"))
                source = source.getParentFile();
        }

        // get the parent directory containing the jar file or the classes
        // directory
        File programDir = source.getParentFile();

        // if we are running inside Eclipse in the target directory, step up to
        // the project level
        if (programDir.getName().equals("target")) programDir = programDir.getParentFile();

        System.setProperty(PROGRAM_DIR, programDir.getAbsolutePath());

        if (!System.getProperties().containsKey(MDL_DIR)) System.setProperty(MDL_DIR, System.getProperty(PROGRAM_DIR));

        if (!System.getProperties().containsKey(LOG_DIR)) System.setProperty(LOG_DIR, System.getProperty(PROGRAM_DIR));

        if (Util.INSTANCE.getDEBUG()) {
            System.out.printf("program.dir = %s%n", System.getProperty(PROGRAM_DIR));
            System.out.printf("mdl.dir = %s%n", System.getProperty(MDL_DIR));
            System.out.printf("log.dir = %s%n", System.getProperty(LOG_DIR));
        }
    }

    public static void main(final String[] args) throws Exception {
        initSystemProperties();
        initSwt();
        startSwtApplication();
    }

    /**
     * Start the application.
     */
    private static void startSwtApplication() throws IllegalAccessException, IllegalArgumentException, InvocationTargetException, ClassNotFoundException,
            NoSuchMethodException, SecurityException {
        // Call main method of SwtMdlBrower via reflection to avoid load time
        // class
        // loading of SWT (which will fail as SWT is not yet on the class path).
        final Class<?> mainDialog = Class.forName("de.treichels.hott.mdlviewer.swt.SwtMdlBrowser");
        final Method main = mainDialog.getMethod("main", String[].class);
        main.invoke(null, new Object[] { new String[] {} });
    }
}

File: HoTT-MdlViewer-SWT/src/main/java/de/treichels/hott/mdlviewer/swt/Launcher.kt
package de.treichels.hott.mdlviewer.swt

import de.treichels.hott.messages.Messages
import de.treichels.hott.util.Util
import java.io.File
import java.net.URL
import java.net.URLClassLoader
import java.util.jar.JarFile

private const val LOG_DIR = "log.dir"
private const val MDL_DIR = "mdl.dir"
private const val PROGRAM_DIR = "program.dir"
private const val PROGRAM_VERSION = "program.version"
private const val MAIN_CLASS_NAME = "de.treichels.hott.mdlviewer.swt.SwtMdlBrowser"

fun main(vararg args: String) {
    Launcher.instance.launch()
}

class Launcher {
    companion object {
        val instance = Launcher()
    }

    // get the location of this class
    private val codeLocation: URL = javaClass.protectionDomain.codeSource.location
    private var sourceFile = File(codeLocation.toURI())
    val programDir: File
    val mdlDir: File
    val logDir: File
    val version: String

    fun launch() {
        // check if swt is already in the classpath
        val classLoader = try {
            Class.forName("org.eclipse.swt.widgets.Composite")
        } catch (e: ClassNotFoundException) {
            // nope - we need to add it

            // determine correct swt jar
            val osName = System.getProperty("os.name").toLowerCase()
            val arch = System.getProperty("os.arch").toLowerCase()

            val name = when {
                osName.startsWith("linux") -> "linux"
                osName.startsWith("windows") -> "windows"
                osName.startsWith("mac") -> "mac"
                else -> osName
            }

            val swtJarName = "swt-$name-$arch.jar"
            val swtDir = File(programDir, "swt")
            val swtJarFile = File(swtDir, swtJarName)
            if (!(swtJarFile.exists() && swtJarFile.isFile && swtJarFile.canRead()))
                throw ClassNotFoundException(Messages.getString("Launcher.SWTNotFound", swtJarFile.absolutePath))

            val swtJarFileURL = swtJarFile.toURI().toURL()
            val systemClassLoader = ClassLoader.getSystemClassLoader()

            // add swt to classpath
            if (systemClassLoader is URLClassLoader) {
                val addMethod = URLClassLoader::class.java.getDeclaredMethod("addURL", URL::class.java)
                addMethod.isAccessible = true
                addMethod.invoke(systemClassLoader, swtJarFileURL)

                systemClassLoader
            } else
                throw UnsupportedOperationException("The system classloader is not a subtype of URLClassLoader. Therefore, we cannot add SWT jar dynamically to the classpath.")
        }

        // Call main method via reflection to avoid load time class loading of SWT (which will fail as SWT is not yet on the class path).
        val mainClass = Class.forName(MAIN_CLASS_NAME)
        val mainMethod = mainClass.getMethod("main", Array<String>::class.java)
        mainMethod.isAccessible = true
        mainMethod.invoke(null, arrayOf<String>())
    }

    /** Initialize system properties. */
    init {
        version = if (sourceFile.name.endsWith(".jar") || sourceFile.name.endsWith(".exe")) {
            // read program version from manifest
            JarFile(sourceFile).use { jarfile ->
                val attributes = jarfile.manifest.mainAttributes
                Messages.getString("Launcher.Version", attributes.getValue("Implementation-Version"), attributes.getValue("Implementation-Build"))
            }
        } else {
            // application was packaged as individual class files, find the classes directory
            while (sourceFile.name != "classes")
                sourceFile = sourceFile.parentFile

            // version is unknown
            System.getProperty(PROGRAM_VERSION, Messages.getString("Launcher.Unknown"))
        }

        // get the parent directory containing the jar file or the classes directory
        val parentFile = sourceFile.parentFile

        // if we are running inside Eclipse in the target directory, step up to the project level
        programDir = if (parentFile.name == "target") parentFile.parentFile else parentFile
        mdlDir = File(System.getProperty(MDL_DIR, programDir.absolutePath))
        logDir = File(System.getProperty(LOG_DIR, programDir.absolutePath))

        System.setProperty(PROGRAM_VERSION, version)
        System.setProperty(PROGRAM_DIR, programDir.absolutePath)
        System.setProperty(MDL_DIR, mdlDir.absolutePath)
        System.setProperty(LOG_DIR, logDir.absolutePath)

        if (Util.DEBUG) {
            System.out.printf("program.dir = %s%n", System.getProperty(PROGRAM_DIR))
            System.out.printf("mdl.dir = %s%n", System.getProperty(MDL_DIR))
            System.out.printf("log.dir = %s%n", System.getProperty(LOG_DIR))
        }
    }
}